### PR TITLE
refactor store to return a reader that only consumes on flush

### PR DESCRIPTION
**Pull request by gvaradarajan**
This refactor was necessary to enable the `DataManager` to preserve data when it fails to upload (upload logic still not included)

### CODE REVIEW

File: micro-rdk/src/common/data_store.rs [hunk 1]
```diff
@@ -164,36 +225,26 @@ impl DataStore for StaticMemoryDataStore {
...
-        Ok(msg_bytes)
-    }
 
     fn from_resource_method_keys(
         collector_keys: Vec<ResourceMethodKey>,
     ) -> Result<Self, DataStoreError> {
         Self::new(collector_keys)
     }
+
+    fn get_reader(
```
Comment (acmorrow) [inline micro-rdk/src/common/data_store.rs:235]: Is it legal for a Producer to put to the underlying `Rb` between two calls to `read_next_message` on the returned `StaticMemoryDataStore`? If so, I think there may be a concern that the position of `head` may be invalidated if old messages need to  be dropped.

File: micro-rdk/src/common/data_store.rs [hunk 2]
```diff
@@ -54,7 +55,18 @@ lazy_static::lazy_static! {
     static ref DATA_STORE_INITIALIZED: AtomicBool = AtomicBool::new(false);
 }
 
+/// A trait for sn entity that is capable of reading from a store without consuming
+/// the messages until a command to flush the read messages is sent
+pub trait DataStoreReader {
```
Comment (acmorrow) [inline micro-rdk/src/common/data_store.rs:60]: If I was a whole lot more proficient with Rust I think I'd be trying to figure out how to make this work with an Iterator that yielded back each `BytesMut`, but as I don't know how to produce such a thing, I don't think it is fair to ask you to do it either ;)

**Review by acmorrow — CHANGES_REQUESTED** [review]
I didn't look particularly closely at the tests, let me know if you would like me to.

File: micro-rdk/src/common/data_store.rs [hunk 3]
```diff
@@ -87,10 +96,50 @@ pub trait DataStore {
...
+        let mut chained = Buf::chain(left, right);
+        chained.advance(self.current_idx - self.start_idx);
+        if !chained.has_remaining() {
+            return Ok(BytesMut::with_capacity(0));
+        }
+        let encoded_len = decode_varint(&mut chained)? as usize;
+        let len_len = length_delimiter_len(encoded_len);
+
+        let msg_bytes = chained.copy_to_bytes(encoded_len);
+        let msg_bytes = BytesMut::from(&msg_bytes[..]);
```
[the remark this answers is not in the document]
Comment (gvaradarajan) [reply micro-rdk/src/common/data_store.rs:134]: I was hoping one, but sadly the answer is two (apparently the implementation of `BytesMut::<&'a [u8]>::from` calls `to_vec` on the u8 slice). I'll look for a more efficient method

File: micro-rdk/src/common/data_store.rs [hunk 4]
```diff
@@ -124,17 +173,28 @@ impl StaticMemoryDataStore {
             .position(|key| key == collector_key)
             .ok_or(DataStoreError::UnknownCollectorKey(collector_key.clone()))
     }
+
+    pub(crate) fn is_buffer_empty(
```
[the remark this answers is not in the document]
Comment (gvaradarajan) [reply micro-rdk/src/common/data_store.rs:177]: no, I wrote it solely for testing. I'll mark it

File: micro-rdk/src/common/data_store.rs (hunk 1 above)
Comment (gvaradarajan) [reply micro-rdk/src/common/data_store.rs:235]: It is not legal, I'll put protections to stop this

File: micro-rdk/src/common/data_store.rs (hunk 2 above)
Comment (gvaradarajan) [reply micro-rdk/src/common/data_store.rs:60]: I'll give it a shot

Comment (gvaradarajan) [reply micro-rdk/src/common/data_store.rs:60]: I have an idea that works, but I want to discuss it with you tomorrow before I move forward

File: micro-rdk/src/common/data_store.rs [hunk 5]
```diff
@@ -54,7 +58,18 @@ lazy_static::lazy_static! {
     static ref DATA_STORE_INITIALIZED: AtomicBool = AtomicBool::new(false);
 }
 
+/// A trait for sn entity that is capable of reading from a store without consuming
```
Comment (acmorrow) [inline micro-rdk/src/common/data_store.rs:61]: typo: `sn`

**Review by acmorrow — APPROVED** [review]
LGTM

**Pull request merged**